Fix failing unlock test + minor modifier improvement

czepluch opened the pull request:
This PR fixes the failing unlock test from issue #198 . The problem was that the wrong party was trying to unlock. The test is still marked as xfails however since the function call that is expected to fail doesn't fail. This will be fixed when issue #96 is taken care of. But now at least the tests are correct. 

Also I've made a small change to a modifier. This is done since the NettingChannelLibrary.sol should not use `msg.sender`

LefterisJP commented:
@czepluch Are you going to further work on this PR? It's been open for quite some time.

czepluch commented:
@LefterisJP yes, I just haven't had the time to do it yet.

czepluch commented:
@hackaugusto You are totally right. Those things shouldn't be changed in the test. I do however think that the problem is related to the fact that https://github.com/czepluch/raiden/blob/cc429cb5bb1bf485bce87c790938b2d01efb1ab5/raiden/tests/smart_contracts/test_netting_channel.py#L785 is `None`.

czepluch commented:
If https://github.com/czepluch/raiden/blob/cc429cb5bb1bf485bce87c790938b2d01efb1ab5/raiden/tests/smart_contracts/test_netting_channel.py#L785 is changed to `our_state` instead of `partner_state`  then `transfer` is indeed a transfer object, but the problem is that the sender of the transfer is node0. The nonce of node0 is 2, and the nonce of the transfer is 2 which means that `updateTransfer` throws. It seems that it's the test that there is something wrong with and not the smart contracts.

hackaugusto commented:
We could simplify the test by splitting it into three:

1. Normal case: A sends B a mediated transfer, A calls `closeSingleTransfer`, B calls unlock.
These are the values of ChannelEnState (I don't think we need to assert this for an unlock test, just giving context):
- channel0.our_state.transfer will be the mediated transfer
- channel0.partner_state.transfer will be None
- channel1.our_state.transfer will be None
- channel1.partner_state will be the mediated transfer
The call to unlock must succed and we need to assert the balances of A and B
*Note*: Additionaly, B should be able to call `closeSingleTransfer` too, but informing A's transfer.

2. Calling unlock twice, same as above, but B calls unlock twice, the second time it must fail, nothing else changes.

3. Calling unlock with an invalid lock. A sends a mediated transfer to B, then the secret is revealed and the lock is unlocked (updating the merkle root), a second message is send from A to B.
The ChannelEndState will be as follows:
- channel0.our_state.transfer will be the **latest** mediated transfer
- channel0.partner_state.transfer will be None
- channel1.our_state.transfer will be None
- channel1.partner_state will be the **latest** mediated transfer
A calls `closeSingleTransfer` and B tries to unlock the *first* lock, this must fail.

czepluch commented:
Updated the PR to now have the correct tests. In order to fix the test I also fixed #96 . 

The smart contracts now keep track of locks that has already been used, thus not allowing the same lock to be unlocked twice.

Furthermore the `unlock` function now checks against the counter party instead of the calling address.
